Query history

applegrain opened the pull request:
Fixes #1103 
This PR makes it possible to search through previous queries on the /graph page.

My approach to this changed once I realized that I could tap into the bootstrap typeahead plugin to keep the UI consistent and my changeset smaller.

**Overview of functionality:**
- unique queries with label-values are stored
- checkbox to toggle historical search on/off
- button to reset the history object in local storage

**Work not included so far and potential tweaks:**
- Ability to opt out of using local storage and use session storage
- keyboard shortcuts to toggle the historical search on/off

![history](https://user-images.githubusercontent.com/10555255/32911781-6627a238-caca-11e7-81e9-06605df44eb7.gif)

applegrain commented on web/ui/static/css/graph.css:
```diff
@@ -6,6 +6,21 @@ body {
+  font-size: 0.8em;
+}
+
+div.query-history button.is-checked {
```
In #3445 I add a similar button, and in whichever PR goes in last I can refactor and reuse the checkbox button code.

applegrain commented:
Hey, let me know what you guys think about the current approach! Happy to make changes as you see fit. 
Ping @fabxc - this resolves an issue you opened a while ago.

Gouthamve commented:
I am not a UI guy, but this feature would be great! I just ran this locally and it would really make me stop copying queries while debugging.

I haven't looked at the code, but this breaks when there are multiple graphs. Just click `Add Graph` at the bottom and then going back to click the toggle button doesn't work. And removing a graph and then toggling an error into the console.

applegrain commented:
Thanks for the feedback @Gouthamve! I'll take a look at those bugs today - I forgot to test with multiple graphs on the page :/

applegrain commented:
@Gouthamve the issue with the multiple graphs has been fixed 👍

Gouthamve commented on web/ui/static/js/graph/index.js:
```diff
@@ -241,48 +251,55 @@ Prometheus.Graph.prototype.populateInsertableMetrics = function() {
+  self.expr.typeahead({
+    source,
+    items: "all",
+    matcher: (item) => {
```
So, the consensus is that arrow functions are not widely supported (On mobile, primarily) and its not worth breaking anyone for a couple of lines of code.

Gouthamve commented:
I tested this locally now, and it works well! :+1: from me, but would like @juliusv or @grobie  to give it a look.

The code also looks a bit cleaned up, nice work!

juliusv commented:
Wow cool. Just testing this now:

- It doesn't seem to store new queries for me when I'm in the search mode? Like, I have to exit the search mode first and re-enable it to make those new queries appear?

- Not getting metric autocompletes in search mode is a bit tough. Ideally it would do both at the same time?

Eeek, I hate to be a pain here, but historically I've (we've?) been extremely careful about adding new UI elements to the `/graph` page, as I don't want it to become too visually overloaded. We even removed some elements over time (auto refresh). IMO every button/input added there has to have a really high value, and only should be added when there's no better way to do the same thing. I think the search functionality is really useful, but I'm on the fence about adding two new buttons for it... is there any way we could just always get both history search + metric autocomplete in the same dropdown? (and then maybe just limit history to a certain number of entries). Or making it visually leaner somehow? Looking at https://kausal.co/blog/first-screencast/, they also have options for this, but somewhat less visually intrusive, under the input field.

applegrain commented:
I totally get your point about keeping the UI minimal, there are definitively ways that I can make it more discrete. We also just got a UX engineer on my team, so I'll see if she has input on how to make it a better experience! Thanks for the feedback 👍

applegrain commented:
I've pushed up changes per your feedback @juliusv 
 
- The big buttons are gone, and now there's just a checkmark toggle. 
- Queries should now be stored when in "history mode".
- Regardless of mode, all metrics show up in the autocomplete dropdown. In "history mode" the recent queries are included as well.

I removed the "clear history" button and instead limited the history to 50 queries (this should probably be higher). I could remove the 50 query limit and add a `clear history` option. 

Queries that return a `200 OK` will be stored. I noticed that we get successful responses even though the query is incomplete at times (for example `http_requ`). I haven't handled this case in the client code since it seems like something that should be taken care of by the server.

Let me know what you think! 

![screen shot 2017-12-20 at 1 04 05 pm](https://user-images.githubusercontent.com/10555255/34226234-49235d22-e586-11e7-95ca-b153e6fcf6a5.png)

juliusv commented on web/ui/static/js/graph/index.js:
```diff
@@ -3,6 +3,15 @@ var graphTemplate;
 
+const keyMap = {
+  enter: 13,
+  r: 18,
```
I think `r` and `c` aren't used anywhere?

applegrain commented:
Thanks for the review! I squashed and fixed the things you brought up. 

- Background color is fixed
- Unused code removed
- Generated bindata.go outside of DEBUG mode

juliusv commented:
:+1: Awesome, thanks! Merging :)
